### ISSUE TITLE
Enable arbitrary nesting of arrow operators

### ISSUE THREAD

**Issue by jzelinskie** [authzed/spicedb#15]
The [arrow operator] can only be used once in an expression (e.g. `org->admin`), but should be able to support traversing an arbitrary number of relations (e.g. `namespace->org->admin`, `tenant->namespace->org->admin->member`).

This would allow for nested arrows in schema

[arrow operator]: https://docs.authzed.com/reference/schema-lang#--arrow

**Comment by jzelinskie** [comment]
I just wanted to add documentation here that you can achieve this today by creating intermediary permissions on objects.

For example:

```
definition token {}
definition user {
	relation access_tokens: token
}

definition team {
	relation member: user

	// Helper for documents to reach the access tokens
	permission member_tokens = member->access_tokens
}

definition document {
    relation writer_teams: team
    relation reader_teams: team

	// This expands to writer_teams->member->access_tokens
    permission edit = writer_teams->member_tokens
    permission view = reader_teams + edit
}
```

**Comment by njuprincerain** [comment]
This can be implemented via introducing a intermediate schema layer. So developers can define multiple hops, like `writer_teams->member->access_tokens`. When spicedb server bootstrap, it could generate the corresponding  intermediate permissions on objects so that the generated schema can work with the current system.

**Comment by birdayz** [comment]
Any plans to implement this?
Without, stuff is really verbose and hard to maintain.

**Comment by josephschorr** [comment]
@birdayz Yes, likely as part of https://github.com/authzed/spicedb/issues/1437, since it will involve schema changes anyway

**Comment by jorihardman-aumni** [comment]
We have a highly hierarchical relationship graph, and many of our most nested entities have permissions that are defined via relationships our users (subjects) have at the top of the hierarchy. In order to facilitate these highly-nested permissions, we need to build out a significant number of proxy permissions that aren't ultimately used by our application logic. Having nested arrows would allow us to only define permissions that we intend to use in applications which would greatly simplify our schema.

The existence of a permission implies that it will be used for domain decisions, and that's often not the case for our schema.

Contrived example:
```
# This is our subject 100% of the time
user {}

org {
  relation member: user
}

nested_entity_1 {
  relation owner_org: org

  # This is a proxy permission - it will not actually be used in application logic.
  permission edit: owner_org->member
}

nested_entity_2 {
  relation owner_entity_1: nested_entity_1

  # This is a proxy permission - it will not actually be used in application logic.
  permission edit: owner_entity_1->edit
}

nested_entity_3 {
  relation owner_entity_2: nested_entity_2
  permission edit: owner_entity_2->edit
}
```

But would be nice to have:
```
# This is our subject 100% of the time
user {}

org {
  relation member: user
}

nested_entity_1 {
  relation owner_org: org
}

nested_entity_2 {
  relation owner_entity_1: nested_entity_1
}

nested_entity_3 {
  relation owner_entity_2: nested_entity_2
  permission edit: owner_entity_2->owner_entity_1->owner_org->member
}
```

**Comment by thackerc** [comment]
> [@birdayz](https://github.com/birdayz) Yes, likely as part of [#1437](https://github.com/authzed/spicedb/issues/1437), since it will involve schema changes anyway

I see there is a preview version of composable schemas. Are you still hoping to include nested arrows at the same time?

**Comment by josephschorr** [comment]
@thackerc We plan to do it; composable schemas has enabled it, but it is not quite scheduled yet

**Comment by thackerc** [comment]
@josephschorr any chance this has been scheduled? Hope you don't mind the nudge.

**Comment by josephschorr** [comment]
@thackerc Not yet; we need to make additional changes to the schema system first, which have been blocked on some legacy systems we're currently fixing

**Comment by tstirrat15** [comment]
We're getting close to being able to start work on this.

I wanted to make sure I understand the use case, though, because I have concerns that nested arrows carry a good amount of implicit behaviors, which could make building out a schema initially easier but then have maintenance ramifications down the road.

For example:

* If you forget or mistype a segment of your path (e.g. you meant to write `foo->bar->baz->qux` but instead wrote `foo->baz->qux`), your permissions would silently fail, because arrowing to a nonexisting relation or permission is technically valid. This is also true of the current state of things, but you're more likely to notice it when you've only got a single arrow to consider.
* If two different definitions are a part of the same chain, and both of those definitions have a permission with the same name, it's possible to unintentionally grant too-broad access. Consider:

```
definition user {}

definition group {
  relation member: user
}

definition organization {
  relation member: user
}

definition folder {
  relation viewer: group | organization
}

definition resource {
  relation parent: folder
  permission view = parent->viewer->member->user
}
```

If the author of the `view` permission is aware of the `organization` part of the `viewer` definition but not the `group` part, that would mean that more users would have the `view` permission than was intended.

Again, this is also a problem in the context of the current system, but a permission being that much more proximate to the relation definitions makes it less likely.

Two things that I can think of that might address pieces of the original intent of this issue:

* If we had `gotoDefinition` support in the SpiceDB language server, it could allow you to quickly traverse a permission path, making development and maintenance of a large, complex, hierarchical schema less painful.
* If we had private permission definitions (e.g. #2733), the question of whether an intermediate permission is a part of the public API falls away.

These are examples; I'd welcome more conversation about alternative approaches or other pain points.

## The Ask
What problem are nested arrows intended to solve? What is painful about working on a schema in the current system?

**Comment by Sese-Schneider** [comment]
@tstirrat15 Here's our concrete use case where nested arrows would meaningfully improve schema clarity.

We have a permission model where a user can be read either by their company or by their specific entity within that `company`, but not both. When a `company_entity` is assigned, only entity-level access should apply:

```
  definition company {
      relation member: user
      permission read = member
  }

  definition company_entity {
      relation company: company
      relation member: user

      // proxy permission — exists solely to enable user.read
      permission read_company = company->read
      permission read = member + company->manage
  }

  definition user {
      relation company: company
      relation company_entity: company_entity

      // company->read is removed when company_entity is set,
      // falling back to company_entity->read
      permission read =
          (company->read - company_entity->read_company) +
          company_entity->read
  }
```

The subtraction `(company->read - company_entity->read_company)` is the key pattern: when a `company_entity` exists, `company_entity->read_company` resolves to the same set as `company->read`, canceling it out.
Without an entity, the subtraction yields an empty set and `company->read` applies fully.

**The problem**: `read_company` on `company_entity` is a pure proxy. It's not used by our application, it's not meaningful on its own, and it exists only because there's no way to express `company_entity->company->read` inline.

With nested arrows, this becomes:

```
  definition company {
      relation member: user
      permission read = member
  }

  definition company_entity {
      relation company: company
      relation member: user

      permission read = member + company->manage
  }

  definition user {
      relation company: company
      relation company_entity: company_entity

      // company->read is removed when company_entity is set,
      // falling back to company_entity->read
      permission read =
          (company->read - company_entity->company->read) +
          company_entity->read
  }
```

The proxy permission disappears entirely, and the subtraction reads naturally: _"company read, minus company read through entity"_; making it immediately clear that the removal only applies when a `company_entity` is set.

That said, #2733 would partially help here by signaling that a permission is internal. But it still requires defining the proxy...
